File: core/src/main/java/io/github/bindglam/neko/registry/EntryScalableRegistry.java
package io.github.bindglam.neko.registry;

import net.kyori.adventure.key.Key;
import org.jetbrains.annotations.NotNull;
import java.util.function.Consumer;

public final class EntryScalableRegistry<T, E extends EntryWritableRegistry.RegistryEntry<T>> extends ScalableRegistry<T> implements EntryWritableRegistry<T, E> {
    private final java.util.function.Supplier<E> entrySupplier;

    public EntryScalableRegistry(@NotNull java.util.function.Supplier<E> entrySupplier) {
        this.entrySupplier = entrySupplier;
    }

    @Override
    public @NotNull T register(@NotNull Key key, @NotNull Consumer<E> entry) {
        if (isLocked) {
            throw new IllegalStateException("The registry is locked");
        }
        if (map.containsKey(key.asString())) {
            throw new IllegalStateException("The registry already contains the key");
        }
        E registryEntry = entrySupplier.get();
        entry.accept(registryEntry);
        T value = registryEntry.toValue();
        map.put(key.asString(), value);
        return value;
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/registry/EntryScalableRegistry.kt
package io.github.bindglam.neko.registry

import net.kyori.adventure.key.Key
import java.util.function.Consumer

class EntryScalableRegistry<T, E : EntryWritableRegistry.RegistryEntry<T>>(private val entrySupplier: () -> E) :
    ScalableRegistry<T>(), EntryWritableRegistry<T, E> {
    override fun register(key: Key, entry: Consumer<E>): T & Any {
        if(isLocked)
            error("The registry is locked")
        if(map.contains(key.asString()))
            error("The registry already contains the key")
        return entrySupplier().apply { entry.accept(this) }.toValue().also { map[key.asString()] = it }
    }
}